src/main/java/com/google/devtools/build/lib/sandbox/SandboxStash.java

```diff
@@ -332,4 +480,145 @@ private static boolean isTestXmlGenerationOrCoverageSpawn(
+      stashContents
+          .symlinksToPathFragment()
+          .keySet()
+          .removeAll(
+              getStringsToRemove(
+                  stashContents.symlinksToPathFragment().keySet(), filesAndSymlinksToKeep));
+    } else {
+      for (var entry : stashContents.dirEntries().entrySet()) {
+        Path absPath = root.getChild(entry.getKey());
+        listContentsRecursively(absPath, timestamp, entry.getValue());
+      }
+    }
+  }
+
+  private static List<String> getStringsToRemove(Set<String> stringSet, Set<String> stringsToKeep) {
+    List<String> stringsToRemove = new ArrayList<>();
+    for (String entry : stringSet) {
+      if (!stringsToKeep.contains(entry)) {
+        stringsToRemove.add(entry);
+      }
+    }
+    return stringsToRemove;
+  }
+
+  private Collection<Path> sortStashesByMatchingTargetSegments(
+      Label target, Collection<Path> stashes) {
+    if (target == null) {
+      return stashes;
+    }
+    List<Path> sortedStashes = new ArrayList<>(stashes);
+    Map<Path, Integer> countMap = new HashMap<>();
+    String[] targetStr = target.getPackageName().split("/");
+    for (Path stash : stashes) {
+      Label stashTarget = sandboxToTarget.getOrDefault(stash, /* defaultValue= */ null);
+      if (stashTarget == null) {
+        sortedStashes.remove(stash);
+        continue;
+      }
+      int count = getMatchingSegmentsCount(targetStr, stashTarget.getPackageName().split("/"));
+      countMap.put(stash, count);
+    }
+    return ImmutableList.sortedCopyOf(
+        Comparator.comparingInt(countMap::get).reversed(), sortedStashes);
+  }
+
+  private int getMatchingSegmentsCount(String[] target, String[] oldTarget) {
+    int count = 0;
+    for (int i = 0; i < target.length; i++) {
+      if (i < oldTarget.length && target[i].equals(oldTarget[i])) {
+        count++;
+      } else {
+        return count;
+      }
+    }
+    return count;
+  }
+
+  private void updateStashContentsAfterRunfilesMove(
+      String stashedRunfiles, String currentRunfiles, StashContents stashContents) {
+    ImmutableList<String> stashedRunfilesSegments =
+        ImmutableList.copyOf(PathFragment.create(stashedRunfiles).segments());
+    StashContents runfilesStashContents =
+        getStashedRunfilesStashContents(stashedRunfilesSegments, 0, stashContents);
+    ImmutableList<String> currentRunfilesSegments =
+        ImmutableList.copyOf(PathFragment.create(currentRunfiles).segments());
+    putStashedRunfilesStashContents(
+        currentRunfilesSegments, 0, stashContents, runfilesStashContents);
+  }
+
+  private StashContents getStashedRunfilesStashContents(
+      List<String> stashedRunfiles, int i, StashContents stashContents) {
+    Preconditions.checkState(i < stashedRunfiles.size());
+    String segment = stashedRunfiles.get(i);
+    Preconditions.checkState(stashContents.dirEntries().containsKey(segment));
+    if (i < stashedRunfiles.size() - 1) {
+      return getStashedRunfilesStashContents(
```

Comment:
I didn't quite get what this function is doing. Is it calling itself recursively in what's essentially a loop? If so, why can't this just be a loop?